teps/0002-custom-tasks.md

```diff
@@ -0,0 +1,392 @@
+
+### Goals
+
+1. Allow non-Pod Task implementations to be built and integrated into "core"
+Tekton Pipelines.
+
+1. Implement a handful of commonly-requested features (e.g., wait, approval)
+as example Custom Tasks in
+[tektoncd/experimental](https://github.com/tektoncd/experimental), to
+demonstrate the contract and act as a working example for future integrators.
+
+1. Provide Go packages and frameworks to help integrators perform common operations required by the Custom Task implementation contract (described below).
+
+### Non-Goals
+
+1. Provide any Custom Task implementations as "official" or "first-party"
+integrations. Some may be added in a future change, but for now the goal
+is just to support _any_ third-party integrations, and let the user install them themselves, or let distributors provide them if they prefer to.
+
+1. As with (1), we don't intend to support any Custom Tasks with any
+special behavior in the `tkn` CLI or the Tekton Dashboard project. As with (1), this may be considered in a future change.
+
+1. Provide any helper libraries or scaffolding for non-Go reconcilers. If sufficient demand for non-Go languages emerges, we may consider helper libraries for them, but at this time we consider this unlikely.
+
+## Requirements
+
+* Add a new CRD type, `Run`, which will be instantiated when a `Pipeline` is run that contains `taskRef`s which have an `apiVersion` that is not `tekton.dev/*` -- `taskRefs` that reference `Task`s and `ClusterTask`s (the only valid values today) will be unaffected.
+
+* Implement and document the Custom Task integration contract (i.e., integrators should update the `Run`'s `.status.conditions` to signal completion).
+
+* Implement and document optional cancellation and timeout behavior of `Run`.
+
+* Provide a package of helper functions and a GitHub template repo to help authors get started.
+
+* Provide sample types and controllers demonstrating simple behavior.
+
+## Proposal
+
+Tekton Pipelines will add a new type, `Run`, initially in `tekton.dev/v1alpha1`.
+
+The `Run` type will take a reference to a user-defined CRD object:
+
+```
+apiVersion: tekton.dev/v1alpha1
+kind: Run
+metadata:
+  generateName: run-
+spec:
+  ref:
+    apiVersion: example.dev/v0
+    kind: Example
+    name: my-example
+```
+
+This references an `Example` CRD type defined by the custom task author, an instance of which is named `my-example`.
+
+When a `Run` object is created, Tekton will validate that the `ref` is specified, and that the specified CRD type is defined, using webhook validation.
+
+After that, Tekton Pipelines expects a custom task author to implement a controller for `Run` objects that reference their type (annotated throughout this proposal with the shorthand `Run<Example>`) to take some action, and eventually update its `.status` to signal completion, either successfully or unsuccessfully, using the `conditions` model used by Tekton PipelineRuns and TaskRuns.
```

Comment:
nit: we are talking about kubernetes status condition and not the tekton *condition* concept (not sure if it's worth being mentionned…)